[RPC] Avoid polluting Python root logger on importing "torch" module

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34871 [RPC] Avoid polluting Python root logger on importing "torch" module**

We used to configure root logger in RPC module. A stream handler is added to `root.handlers`. This is not desired behavior for pytorch users. We should instead keep the root logger handler list untouched.

We can configure the logger local to the rpc module, set it's log level, so it doesn't use it's ancestor, which is usually the root which has no stream handlers in most cases.
https://docs.python.org/3/library/logging.html#logging.Logger.setLevel

And add a stream handler to make it output to stdout, even if the root handlers is not configured and has an empty list.
https://docs.python.org/3/library/logging.html#logging.Logger.addHandler
https://docs.python.org/3/library/logging.handlers.html#logging.StreamHandler

Differential Revision: [D7677493](https://our.internmc.facebook.com/intern/diff/D7677493/)

xush6528 committed:
Avoid polute Python root logger because of importing pytorch module

We used to configure root logger in RPC module. A stream handler is added to `root.handlers`. This is not desired behavior for pytorch users. We should instead keep the root logger handler list untouched.

We can configure the logger local to the rpc module, set it's log level, so it doesn't use it's ancestor, which is usually the root which has no stream handlers in most cases.
https://docs.python.org/3/library/logging.html#logging.Logger.setLevel

And add a stream handler to make it output to stdout, even if the root handlers is not configured and has an empty list.
https://docs.python.org/3/library/logging.html#logging.Logger.addHandler
https://docs.python.org/3/library/logging.handlers.html#logging.StreamHandler

Differential Revision: [D7677493](https://our.internmc.facebook.com/intern/diff/D7677493/)

[ghstack-poisoned]

ezyang commented on torch/distributed/rpc/api.py:
```diff
@@ -37,8 +37,10 @@
-logging.basicConfig()
 logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+logger.addHandler(logging.StreamHandler(sys.stdout))
```
This still looks wrong. The user should control all aspects of how logging occurs; in general, libraries should not setup any handlers. Citations: https://docs.python-guide.org/writing/logging/#logging-in-a-library https://docs.python.org/3/howto/logging.html#configuring-logging-for-a-library

ezyang commented:
I'm not going to block this PR as it is a strict improvement, but I still think that setting a handler here is very questionable.

xush6528 committed:
Update on "[RPC] Avoid polluting Python root logger on importing "torch" module"


We used to configure root logger in RPC module. A stream handler is added to `root.handlers`. This is not desired behavior for pytorch users. We should instead keep the root logger handler list untouched.

We can configure the logger local to the rpc module, set it's log level, so it doesn't use it's ancestor, which is usually the root which has no stream handlers in most cases.
https://docs.python.org/3/library/logging.html#logging.Logger.setLevel

And add a stream handler to make it output to stdout, even if the root handlers is not configured and has an empty list.
https://docs.python.org/3/library/logging.html#logging.Logger.addHandler
https://docs.python.org/3/library/logging.handlers.html#logging.StreamHandler

Differential Revision: [D7677493](https://our.internmc.facebook.com/intern/diff/D7677493/)

[ghstack-poisoned]

xush6528 commented:
@ezyang 
Thanks for the suggestion. 
I will leave this local logger untouched, so it can get the config of the root logger. 
If users never configure the root logger (usually by `logging.basicConfig()`), the `logging.lastResort` logger will be used, which is set to warning level. 
It's sufficient for RPC module, because we only log error messages.
